addons/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/ws/IhcClient.java

```diff
@@ -0,0 +1,558 @@
+     * Enable resources runtime value notifications.
+     *
+     * @param resourceIdList
+     *            List of resource Identifiers.
+     * @return True is connection successfully opened.
+     */
+    public synchronized void enableRuntimeValueNotifications(Set<Integer> resourceIdList) throws IhcExecption {
+
+        resourceInteractionService.enableRuntimeValueNotifications(resourceIdList);
+    }
+
+    /**
+     * Wait runtime value notifications.
+     *
+     * Runtime value notification should firstly be activated by
+     * enableRuntimeValueNotifications function.
+     *
+     * @param timeoutInSeconds
+     *            How many seconds to wait notifications.
+     * @return List of received runtime value notifications.
+     * @throws SocketTimeoutException
+     */
+    private List<WSResourceValue> waitResourceValueNotifications(int timeoutInSeconds)
+            throws IhcExecption, SocketTimeoutException {
+
+        List<WSResourceValue> list = resourceInteractionService.waitResourceValueNotifications(timeoutInSeconds);
+
+        for (WSResourceValue val : list) {
+            resourceValues.put(val.getResourceID(), val);
+        }
+        return list;
+    }
+
+    /**
+     * Query resource value from controller.
+     *
+     *
+     * @param resoureId Resource Identifier.
+     * @return Resource value.
+     */
+    public WSResourceValue resourceQuery(int resoureId) throws IhcExecption {
+        return resourceInteractionService.resourceQuery(resoureId);
+    }
+
+    /**
+     * Get resource value information.
+     *
+     * Function return resource value from internal memory, if data is not
+     * available information is read from the controller.
+     *
+     * Resource value's value field (e.g. floatingPointValue) could be old
+     * information.
+     *
+     * @param resoureId Resource Identifier.
+     * @return Resource value.
+     */
+    public WSResourceValue getResourceValueInformation(int resourceId) throws IhcExecption {
+        if (resourceId != 0) {
+            WSResourceValue data = resourceValues.get(resourceId);
+            if (data == null) {
+                // data is not available, read it from the controller
+                data = resourceInteractionService.resourceQuery(resourceId);
+            }
+            return data;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Update resource value to controller.
+     *
+     *
+     * @param value Resource value.
+     * @return True if value is successfully updated.
+     */
+    public boolean resourceUpdate(WSResourceValue value) throws IhcExecption {
+        return resourceInteractionService.resourceUpdate(value);
+    }
+
+    /**
+     * The IhcReader runs as a separate thread.
+     *
+     * Thread listen resource value notifications from IHC / ELKO LS controller.
+     */
+    private class IhcResourceValueNotificationListener extends Thread {
+        private boolean interrupted = false;
```

Comment:
I believe this field should be `volatile` as it's set by different threads.

Reply:
Done